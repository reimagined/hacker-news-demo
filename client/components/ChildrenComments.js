import React from "react";

import Comment from "./Comment";

const ChildrenComments = ({ replies, level, comments, users }) => {
  if (!replies.length) {
    return null;
  }

  const currentLevel = level ? level + 1 : 1;
  return (
    <div>
      {replies.map(replyId => {
        const comment = comments.find(({ id }) => id === replyId);
        const { id, text, replies, createdAt, createdBy } = comment;
        const user = users.find(user => user.id === createdBy);

        return (
          <Comment
            key={id}
            level={currentLevel}
            id={id}
            content={text}
            user={user.name}
            date={new Date(createdAt)}
            showReply
          >
            <ChildrenComments
              replies={replies}
              level={currentLevel}
              comments={comments}
              users={users}
            />
          </Comment>
        );
      })}
    </div>
  );
};

export default ChildrenComments;
