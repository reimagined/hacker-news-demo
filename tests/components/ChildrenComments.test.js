import React from 'react'
import { shallow } from 'enzyme'

import ChildrenComments from '../../client/components/ChildrenComments'

it('ChildrenComments renders correctly', () => {
  const story = {
    id: 'story-id',
    comments: [
      {
        id: 'comment-id',
        parentId: 'story-id',
        text: 'comment',
        createdAt: new Date(0),
        createdBy: 'user-id',
        createdByName: 'SomeUser'
      },
      {
        id: 'reply-id',
        parentId: 'comment-id',
        text: 'reply',
        createdAt: new Date(0),
        createdBy: 'user-id',
        createdByName: 'SomeUser'
      }
    ]
  }

  const markup = shallow(
    <ChildrenComments
      storyId={story.id}
      parentId={story.id}
      comments={story.comments}
      level={0}
    />
  )
  expect(markup).toMatchSnapshot()
})

it('Empty renders correctly', () => {
  const story = {
    id: 'story-id',
    comments: []
  }

  const markup = shallow(
    <ChildrenComments
      storyId={story.id}
      parentId={story.id}
      comments={story.comments}
      level={0}
    />
  )
  expect(markup).toMatchSnapshot()
})
