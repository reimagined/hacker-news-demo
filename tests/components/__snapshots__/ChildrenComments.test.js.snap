// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChildrenComments renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <Comment
        createdAt={1970-01-01T00:00:00.000Z}
        createdBy="user-id"
        createdByName="SomeUser"
        id="comment-id"
        level={1}
        parentId="story-id"
        storyId="story-id"
        text="comment"
    >
        <ReplyLink
            commentId="comment-id"
            level={1}
            storyId="story-id"
        />
        <ChildrenComments
            comments={
                Array [
                    Object {
                      "createdAt": 1970-01-01T00:00:00.000Z,
                      "createdBy": "user-id",
                      "createdByName": "SomeUser",
                      "id": "comment-id",
                      "parentId": "story-id",
                      "text": "comment",
                    },
                    Object {
                      "createdAt": 1970-01-01T00:00:00.000Z,
                      "createdBy": "user-id",
                      "createdByName": "SomeUser",
                      "id": "reply-id",
                      "parentId": "comment-id",
                      "text": "reply",
                    },
                  ]
            }
            level={1}
            parentId="comment-id"
            storyId="story-id"
        />
    </Comment>
</div>,
  "nodes": Array [
    <div>
      <Comment
            createdAt={1970-01-01T00:00:00.000Z}
            createdBy="user-id"
            createdByName="SomeUser"
            id="comment-id"
            level={1}
            parentId="story-id"
            storyId="story-id"
            text="comment"
      >
            <ReplyLink
                  commentId="comment-id"
                  level={1}
                  storyId="story-id"
            />
            <ChildrenComments
                  comments={
                        Array [
                              Object {
                                "createdAt": 1970-01-01T00:00:00.000Z,
                                "createdBy": "user-id",
                                "createdByName": "SomeUser",
                                "id": "comment-id",
                                "parentId": "story-id",
                                "text": "comment",
                              },
                              Object {
                                "createdAt": 1970-01-01T00:00:00.000Z,
                                "createdBy": "user-id",
                                "createdByName": "SomeUser",
                                "id": "reply-id",
                                "parentId": "comment-id",
                                "text": "reply",
                              },
                            ]
                  }
                  level={1}
                  parentId="comment-id"
                  storyId="story-id"
            />
      </Comment>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <ChildrenComments
        comments={
                Array [
                        Object {
                          "createdAt": 1970-01-01T00:00:00.000Z,
                          "createdBy": "user-id",
                          "createdByName": "SomeUser",
                          "id": "comment-id",
                          "parentId": "story-id",
                          "text": "comment",
                        },
                        Object {
                          "createdAt": 1970-01-01T00:00:00.000Z,
                          "createdBy": "user-id",
                          "createdByName": "SomeUser",
                          "id": "reply-id",
                          "parentId": "comment-id",
                          "text": "reply",
                        },
                      ]
        }
        level={0}
        parentId="story-id"
        storyId="story-id"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "comments": Array [
            Object {
              "createdAt": 1970-01-01T00:00:00.000Z,
              "createdBy": "user-id",
              "createdByName": "SomeUser",
              "id": "comment-id",
              "parentId": "story-id",
              "text": "comment",
            },
            Object {
              "createdAt": 1970-01-01T00:00:00.000Z,
              "createdBy": "user-id",
              "createdByName": "SomeUser",
              "id": "reply-id",
              "parentId": "comment-id",
              "text": "reply",
            },
          ],
          "level": 0,
          "parentId": "story-id",
          "storyId": "story-id",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <Comment
                    createdAt={1970-01-01T00:00:00.000Z}
                    createdBy="user-id"
                    createdByName="SomeUser"
                    id="comment-id"
                    level={1}
                    parentId="story-id"
                    storyId="story-id"
                    text="comment"
          >
                    <ReplyLink
                              commentId="comment-id"
                              level={1}
                              storyId="story-id"
                    />
                    <ChildrenComments
                              comments={
                                        Array [
                                                  Object {
                                                    "createdAt": 1970-01-01T00:00:00.000Z,
                                                    "createdBy": "user-id",
                                                    "createdByName": "SomeUser",
                                                    "id": "comment-id",
                                                    "parentId": "story-id",
                                                    "text": "comment",
                                                  },
                                                  Object {
                                                    "createdAt": 1970-01-01T00:00:00.000Z,
                                                    "createdBy": "user-id",
                                                    "createdByName": "SomeUser",
                                                    "id": "reply-id",
                                                    "parentId": "comment-id",
                                                    "text": "reply",
                                                  },
                                                ]
                              }
                              level={1}
                              parentId="comment-id"
                              storyId="story-id"
                    />
          </Comment>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <Comment
                    createdAt={1970-01-01T00:00:00.000Z}
                    createdBy="user-id"
                    createdByName="SomeUser"
                    id="comment-id"
                    level={1}
                    parentId="story-id"
                    storyId="story-id"
                    text="comment"
          >
                    <ReplyLink
                              commentId="comment-id"
                              level={1}
                              storyId="story-id"
                    />
                    <ChildrenComments
                              comments={
                                        Array [
                                                  Object {
                                                    "createdAt": 1970-01-01T00:00:00.000Z,
                                                    "createdBy": "user-id",
                                                    "createdByName": "SomeUser",
                                                    "id": "comment-id",
                                                    "parentId": "story-id",
                                                    "text": "comment",
                                                  },
                                                  Object {
                                                    "createdAt": 1970-01-01T00:00:00.000Z,
                                                    "createdBy": "user-id",
                                                    "createdByName": "SomeUser",
                                                    "id": "reply-id",
                                                    "parentId": "comment-id",
                                                    "text": "reply",
                                                  },
                                                ]
                              }
                              level={1}
                              parentId="comment-id"
                              storyId="story-id"
                    />
          </Comment>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ChildrenComments
    comments={
        Array [
            Object {
              "createdAt": 1970-01-01T00:00:00.000Z,
              "createdBy": "user-id",
              "createdByName": "SomeUser",
              "id": "comment-id",
              "parentId": "story-id",
              "text": "comment",
            },
            Object {
              "createdAt": 1970-01-01T00:00:00.000Z,
              "createdBy": "user-id",
              "createdByName": "SomeUser",
              "id": "reply-id",
              "parentId": "comment-id",
              "text": "reply",
            },
          ]
    }
    level={0}
    parentId="story-id"
    storyId="story-id"
/>,
}
`;

exports[`ChildrenComments renders correctly 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <Comment
        createdAt={1970-01-01T00:00:00.000Z}
        createdBy="user-id"
        createdByName="SomeUser"
        id="reply-id"
        level={2}
        parentId="comment-id"
        storyId="story-id"
        text="reply"
    >
        <ReplyLink
            commentId="reply-id"
            level={2}
            storyId="story-id"
        />
        <ChildrenComments
            comments={
                Array [
                    Object {
                      "createdAt": 1970-01-01T00:00:00.000Z,
                      "createdBy": "user-id",
                      "createdByName": "SomeUser",
                      "id": "comment-id",
                      "parentId": "story-id",
                      "text": "comment",
                    },
                    Object {
                      "createdAt": 1970-01-01T00:00:00.000Z,
                      "createdBy": "user-id",
                      "createdByName": "SomeUser",
                      "id": "reply-id",
                      "parentId": "comment-id",
                      "text": "reply",
                    },
                  ]
            }
            level={2}
            parentId="reply-id"
            storyId="story-id"
        />
    </Comment>
</div>,
  "nodes": Array [
    <div>
      <Comment
            createdAt={1970-01-01T00:00:00.000Z}
            createdBy="user-id"
            createdByName="SomeUser"
            id="reply-id"
            level={2}
            parentId="comment-id"
            storyId="story-id"
            text="reply"
      >
            <ReplyLink
                  commentId="reply-id"
                  level={2}
                  storyId="story-id"
            />
            <ChildrenComments
                  comments={
                        Array [
                              Object {
                                "createdAt": 1970-01-01T00:00:00.000Z,
                                "createdBy": "user-id",
                                "createdByName": "SomeUser",
                                "id": "comment-id",
                                "parentId": "story-id",
                                "text": "comment",
                              },
                              Object {
                                "createdAt": 1970-01-01T00:00:00.000Z,
                                "createdBy": "user-id",
                                "createdByName": "SomeUser",
                                "id": "reply-id",
                                "parentId": "comment-id",
                                "text": "reply",
                              },
                            ]
                  }
                  level={2}
                  parentId="reply-id"
                  storyId="story-id"
            />
      </Comment>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <ChildrenComments
        comments={
                Array [
                        Object {
                          "createdAt": 1970-01-01T00:00:00.000Z,
                          "createdBy": "user-id",
                          "createdByName": "SomeUser",
                          "id": "comment-id",
                          "parentId": "story-id",
                          "text": "comment",
                        },
                        Object {
                          "createdAt": 1970-01-01T00:00:00.000Z,
                          "createdBy": "user-id",
                          "createdByName": "SomeUser",
                          "id": "reply-id",
                          "parentId": "comment-id",
                          "text": "reply",
                        },
                      ]
        }
        level={1}
        parentId="comment-id"
        storyId="story-id"
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "comments": Array [
            Object {
              "createdAt": 1970-01-01T00:00:00.000Z,
              "createdBy": "user-id",
              "createdByName": "SomeUser",
              "id": "comment-id",
              "parentId": "story-id",
              "text": "comment",
            },
            Object {
              "createdAt": 1970-01-01T00:00:00.000Z,
              "createdBy": "user-id",
              "createdByName": "SomeUser",
              "id": "reply-id",
              "parentId": "comment-id",
              "text": "reply",
            },
          ],
          "level": 1,
          "parentId": "comment-id",
          "storyId": "story-id",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <Comment
                    createdAt={1970-01-01T00:00:00.000Z}
                    createdBy="user-id"
                    createdByName="SomeUser"
                    id="reply-id"
                    level={2}
                    parentId="comment-id"
                    storyId="story-id"
                    text="reply"
          >
                    <ReplyLink
                              commentId="reply-id"
                              level={2}
                              storyId="story-id"
                    />
                    <ChildrenComments
                              comments={
                                        Array [
                                                  Object {
                                                    "createdAt": 1970-01-01T00:00:00.000Z,
                                                    "createdBy": "user-id",
                                                    "createdByName": "SomeUser",
                                                    "id": "comment-id",
                                                    "parentId": "story-id",
                                                    "text": "comment",
                                                  },
                                                  Object {
                                                    "createdAt": 1970-01-01T00:00:00.000Z,
                                                    "createdBy": "user-id",
                                                    "createdByName": "SomeUser",
                                                    "id": "reply-id",
                                                    "parentId": "comment-id",
                                                    "text": "reply",
                                                  },
                                                ]
                              }
                              level={2}
                              parentId="reply-id"
                              storyId="story-id"
                    />
          </Comment>
</div>,
        "_debugID": 4,
        "_renderedOutput": <div>
          <Comment
                    createdAt={1970-01-01T00:00:00.000Z}
                    createdBy="user-id"
                    createdByName="SomeUser"
                    id="reply-id"
                    level={2}
                    parentId="comment-id"
                    storyId="story-id"
                    text="reply"
          >
                    <ReplyLink
                              commentId="reply-id"
                              level={2}
                              storyId="story-id"
                    />
                    <ChildrenComments
                              comments={
                                        Array [
                                                  Object {
                                                    "createdAt": 1970-01-01T00:00:00.000Z,
                                                    "createdBy": "user-id",
                                                    "createdByName": "SomeUser",
                                                    "id": "comment-id",
                                                    "parentId": "story-id",
                                                    "text": "comment",
                                                  },
                                                  Object {
                                                    "createdAt": 1970-01-01T00:00:00.000Z,
                                                    "createdBy": "user-id",
                                                    "createdByName": "SomeUser",
                                                    "id": "reply-id",
                                                    "parentId": "comment-id",
                                                    "text": "reply",
                                                  },
                                                ]
                              }
                              level={2}
                              parentId="reply-id"
                              storyId="story-id"
                    />
          </Comment>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ChildrenComments
    comments={
        Array [
            Object {
              "createdAt": 1970-01-01T00:00:00.000Z,
              "createdBy": "user-id",
              "createdByName": "SomeUser",
              "id": "comment-id",
              "parentId": "story-id",
              "text": "comment",
            },
            Object {
              "createdAt": 1970-01-01T00:00:00.000Z,
              "createdBy": "user-id",
              "createdByName": "SomeUser",
              "id": "reply-id",
              "parentId": "comment-id",
              "text": "reply",
            },
          ]
    }
    level={1}
    parentId="comment-id"
    storyId="story-id"
/>,
}
`;

exports[`Empty renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": null,
  "nodes": Array [
    null,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <ChildrenComments
        comments={Array []}
        level={0}
        parentId="story-id"
        storyId="story-id"
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "comments": Array [],
          "level": 0,
          "parentId": "story-id",
          "storyId": "story-id",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": null,
        "_debugID": 6,
        "_renderedOutput": null,
      },
      "_renderedNodeType": 2,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ChildrenComments
    comments={Array []}
    level={0}
    parentId="story-id"
    storyId="story-id"
/>,
}
`;
